### PR TITLE
Randomise the ordering of the C unit tests.

### PR DESCRIPTION
**Pull request by pauli-oracle**
- [x] tests are added or updated

This PR provides the capability to randomise the ordering of the C unit tests.  This can be used to help detect inter-test dependencies.

By default, the order randomisation is disabled.  It can be turned on by setting the `OPENSSL_TEST_RAND_ORDER` environment variable.  If this environment variable is set to a positive numeric value, that value is used as the seed (which allows test runs to be accurately repeated). Otherwise, the current time is used as a seed.  In either case, the seed used is printed out at the top of the test's run.

The `bioprinttest.c` has inter-test dependencies which have been addressed as well.

### CODE REVIEW

**Comment by pauli-oracle** [comment]
The Travis failures seem to be unrelated to this change.

The big question is should the randomised tests be made default (& explicitly disablable) rather than disabled but enableable?  There is some small potential for confusion from the output as a downside.  The upside is the knowledge that the tests don't interact with each other (at least to some degree).

This will become more important when more extensive interop testing is undertaken.

File: test/testutil/driver.c [hunk 1]
```diff
@@ -165,8 +198,19 @@ int run_tests(const char *test_prog_name)
...
+            else
+                do
+                    jstep = rand() % all_tests[i].num;
+                while (jstep == 0 || gcd(all_tests[i].num, jstep) != 1);
+
+            for (jj = 0; jj < all_tests[i].num; jj++) {
+                int ret;
+
+                j = (j + jstep) % all_tests[i].num;
+                ret = all_tests[i].param_test_fn(j);
```
Comment (levitte) [inline test/testutil/driver.c]: I think the TAP output below (new line 229) should change to this:

``` C
                     test_printf_stdout("%*s%s %d - iteration %d\n", level, "", verdict, jj, j + 1);
```

File: test/testutil/driver.c [hunk 2]
```diff
@@ -137,7 +159,18 @@ int run_tests(const char *test_prog_name)
...
+    if (seed != 0)
+        for (i = num_tests - 1; i >= 1; i--) {
+            j = rand() % (1 + i);
+            ii = permute[j];
+            permute[j] = permute[i];
+            permute[i] = ii;
+        }
+
+    for (ii = 0; ii != num_tests; ++ii) {
+        i = permute[ii];
```
Comment (levitte) [inline test/testutil/driver.c]: In tha TAP output at new line 185, change the argument `i + 1` to `ii + 1`

**Comment by richsalz** [comment]
I think the default really has to be repeatable output and therefore non-random testing.
But a CI build that turns on randomization is very worthwhile.

**Comment by pauli-oracle** [comment]
I've updated the output and added randomised ordering to one of the Travis tests.

**Comment by levitte** [comment]
@pauli-oracle, if you want to have that Travis build job exercised, please add the following line to your last commit message:

```
[extended tests]
```

**Comment by pauli-oracle** [comment]
Hmmmm, I must have missed something with the Travis config.

I'll make the change in the morning.

**Comment by levitte** [comment]
> Hmmmm, I must have missed something with the Travis config.

The jobs with `EXTENDED_TESTS=yes` will only actually be performed with the last commit message contains `[extended tests]`.

> I'll make the change in the morning.

Is it morning yet?  ;-)

**Comment by t-j-h** [comment]
In @pauli-oracle's timezone it is pre-6am so from his view not yet "morning".

**Comment by levitte** [comment]
Okie

**Comment by pauli-oracle** [comment]
I've made the update.  I perhaps should have modified a different test so the randomised tests always run.

Tim, your up early :)

**Comment by richsalz** [comment]
reconfirm

**Comment by levitte** [comment]
Merged.  5584fd7f3d306ea0f610be9d69e64050421b2a0d and 4861933a135658a85eaaab80ce2133af4a9d413c

**Pull request closed without merging**